src/Shared/TagTransformerJsonHelper.cs

```diff
@@ -0,0 +1,64 @@
+// <copyright file="TagTransformerJsonHelper.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System.Text.Json;
+#if NET6_0_OR_GREATER
+using System.Text.Json.Serialization;
+#endif
+
+namespace OpenTelemetry.Internal;
+
+/// <summary>
+/// This class has to be partial so that JSON source generator can provide code for the JsonSerializerContext.
+/// </summary>
+#pragma warning disable SA1601 // Partial elements should be documented
+internal static partial class TagTransformerJsonHelper
+{
+#if NET6_0_OR_GREATER
+    // In net6.0 or higher ships System.Text.Json "in box",
+    // meaning the consumer automatically got upgraded to use v6.0 System.Text.Json which uses source generator.
```

Comment:
Does the user automatically get source generation?

Reply:
Sorry, I probably don't understand the question.

The `ArrayTagJsonContext` is annotated with the `JsonSerializable` attributes - this is what the source generator in `System.Text.Json 6.0.0` will pick up on, it will then generate methods (and properties) into the `ArrayTagJsonContext`. All of those will be exposed via the `Default` property. So here we pass that property to the new `Serialize` overload which uses it to do the actual work, instead of relying on reflection.

Reply:
My concern was with the statement in comment:

> meaning the consumer automatically got upgraded to use v6.0 System.Text.Json which uses source generator.

This sentence sounded like we didn't have to do define anything and the `System.Text.Json` uses source generator by default.

Reply:
Rephrasing suggestion for the comment:
" On .NET 6 and above, v6+ of `System.Text.Json` ships “in the box” as part of the base class libraries. This allows us to use source generator on higher targets."

Reply:
I see - yes, we should reword the comment to make it clearer.

Reply:
Found this blog helpful for me: https://devblogs.microsoft.com/dotnet/try-the-new-system-text-json-source-generator/#introducing-jsontypeinfot-jsontypeinfo-and-jsonserializercontext
And yes, comment is a bit confusing, will update it.

"Implementations of the JsonTypeInfo<T>, JsonTypeInfo, and JsonSerializerContext types are the primary result of JSON source generation.

The JsonTypeInfo<T> type contains structured information about how to serialize and deserialize a single type. This information can contain metadata about how to access its members. This information is required when the serializer itself is performing the (de)serialization of the type, using the robust logic it has to support all the features that can be configured with JsonSerializerOptions or serialization attributes. "

Reply:
> use v6.0 System.Text.Json which uses source generator.

This part of the sentence is what sounds misleading. v6.0 of `System.Text.Json` does not always use source generator. v6.0 of the package just lets us use source generator. We have to explicitly provide the custom `JsonSerializerContext` to make it use source generator.

Reply:
Updated the comment to reduce the above mentioning confusion.